messagecatalog/spec.md

```diff
@@ -0,0 +1,1475 @@
+related either by ownership or by a shared subject matter context.
+
+Implementations of this specification MAY associate access control rules with
+message definition groups. For instance, a user or a group of users might be
+given write access only to a particular message definition group that their
+organization owns. If trade secret protection is necessary for an application or
+parts of it and the definition structure would give some of those away, read
+access to a group of definitions might likewise be restricted.
+
+This specification does not define management constructs for access control
+rules.
+
+### 2.1.1. Message definition group attributes
+
+The data model for a message definition group consists of these attributes:
+
+#### `id` (message definition group)
+
+- Type: `String`
+- Description: Identifies the message definition group.
+- Constraints:
+  - REQUIRED. IMMUTABLE.
+  - MUST be a non-empty string
+  - MUST conform with RFC3986/3.3 `segment-nz-nc` syntax. This allows for "dot
+    notation", e.g. `org.example.myapp.module` for logical organization of
+    message definition groups, even though the namespace is flat.
+  - MUST be unique within the scope of the message catalog
+- Examples:
+  - mygroup
+  - my-group
+  - org.example.myapp.module
+  - events\@example.com
+
+#### `specversion` (message definition group)
+
+- Type: `String`
+- Description: Version of this specification that this document complies with.
+- Constraints:
+  - REQUIRED.
+  - MUST be "0.1" for this version of the specification.
+
+#### `format` (message definition group)
+
+- Type: `String`
+- Description: Defines the message definition format managed by this message
+  definition group. The format MAY be one of formats defined in this
+  specification (CloudEvents, NATS, AMQP, MQTT) or a custom format. All
+  definitions in this group MUST conform with the rules of the chosen format.
+- Constraints:
+  - REQUIRED. IMMUTABLE.
+  - MUST NOT be modified if at least one definition exists in the group.
+  - "CloudEvents", "NATS", "AMQP", "MQTT" for the definition formats defined in
+    this specification, or another format defined elsewhere. If one of the names
+    of definition formats defined in this specification is used, those
+    definitions MUST be conformant with this specification.
+- Examples:
+  - "CloudsEvents"
+  - "AMQP"
+  - "MQTT"
+
+#### `description` (message definition group)
+
+- Type: `String`
+- Description: Concisely explains the purpose of the message definition group.
+- Constraints:
+  - OPTIONAL
+- Examples:
+  - "This group holds Message definitions for the fabulous example app."
+
+#### `documentation` (message definition group)
+
+- Type: `String`
+- Description: Documents the purpose of the message definition group in greater
+  detail than the `description`. The text MAY use [CommonMark](https://commonmark.org/) markdown formatting.
+- Constraints:
+  - OPTIONAL
+
+#### `tags` (message definition group)
```

Comment:
My initial reaction was to make these name/value pairs, but then I realized we could get the same results via single-string-tags IF we also restrict the character set. For example, don't allow `,` in name or value (at least not w/o an escape), but do allow `=`.